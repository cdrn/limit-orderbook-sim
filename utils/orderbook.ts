export interface Order extends OrderSubmissionInterface {
  idNumber: string; // Monotonically increasing order ID
  entryTime: bigint | null;
  eventTime: bigint;
  nextOrder: Order | null;
  prevOrder: Order | null;
  parentLimit: Limit | null;
}

export interface OrderSubmissionInterface {
  buyOrSell: boolean; // Buy: true, sell: false
  shares: bigint;
  limit: bigint;
}

/*
 * This object represents a singe limit. These should be organised as a binary tree of limits by price, which allows us
 * to find our desired limit in O(log(n)) time. It also contains a reference to a linked list of Orders at that price.
 */
export interface Limit {
  limitPrice: bigint;
  size: bigint;
  totalVolume: bigint;
  parent: Limit | null;
  leftChild: Limit | null;
  rightChild: Limit | null;
  // Limits must always contain at least one order -- even if the head and tail are the same.
  headOrder: Order;
  tailOrder: Order;
}

/**
 * The entire orderbook for a given asset pair. This is a binary tree of Limits, which are linked lists of Orders.
 */
export interface Book {
  buyTree: Limit | null;
  sellTree: Limit | null;
  lowestSell: Limit | null;
  highestBuy: Limit | null;
  orderMap: Record<string, Order>;
  limitMap: Record<string, Limit>;
}

//////////////////////
// HELPER FUNCTIONS //
//////////////////////
/**
 * Traverse the binary tree of Limits, and return the Limit node with the most similar price to
 * the given price.
 * @param price
 * @param tree
 * @returns
 */
export const findClosestLimitFromPrice = (
  price: bigint,
  tree: Limit | null
): Limit | null => {
  if (tree === null) {
    return null;
  }

  if (tree.limitPrice && price < tree.limitPrice) {
    return findClosestLimitFromPrice(price, tree?.leftChild);
  } else if (tree.limitPrice && price < tree.limitPrice) {
    return findClosestLimitFromPrice(price, tree?.rightChild);
  } else {
    return tree;
  }
};

/**
 * Create a limit order where one previously did not exist.
 * @param orderbook
 * @param order
 */
export const createGenesisLimit = (order: Order) => {
  const limit = {
    limitPrice: order.limit,
    size: order.shares,
    totalVolume: order.shares,
    parent: null,
    leftChild: null,
    rightChild: null,
    headOrder: order,
    tailOrder: order,
  };
  return limit;
};

/**
 * Add an order to an existing "limit" structure.
 * @param orderbook
 * @param order
 */
export const addOrderToLimit = (BuyOrSellTree: Limit, order: Order) => {
  return;
};

export const addOrderToOrderbook = (orderbook: Book, order: Order): Book => {
  // If theres no limit at the order root, we have to create one
  // Check separately for buy and sell limits
  if (order.buyOrSell) {
    if (orderbook.buyTree === null) {
      const genesisLimit = createGenesisLimit(order);
      orderbook.buyTree = genesisLimit;
      orderbook.limitMap[String(genesisLimit.limitPrice)] = genesisLimit;
      console.log("THE GENESIS BUY ORDER IS", orderbook.buyTree);
    } else {
      addOrderToLimit(orderbook.buyTree, order);
    }
  } else if (!order.buyOrSell) {
    if (orderbook.sellTree === null) {
      const genesisLimit = createGenesisLimit(order);
      orderbook.sellTree = genesisLimit;
      orderbook.limitMap[String(genesisLimit.limitPrice)] = genesisLimit;
      orderbook.sellTree = createGenesisLimit(order);
      console.log("THE GENESIS SELL ORDER IS", orderbook.sellTree);
    } else {
      addOrderToLimit(orderbook.sellTree, order);
    }
  }
  // Always add the order to the orderMap
  orderbook.orderMap[order.idNumber] = order;
};

export const getLastOrderId = (orderbook: Book) => {
  return (
    Object.keys(orderbook.orderMap).sort((a, b) =>
      Number(BigInt(b) - BigInt(a))
    )[0] || 0
  ); // TODO: this is inefficient lol
};

// This will be the orderbook for a given asset pair.
export const FastLimitOrderbook = () => {
  // Represents the entire orderbook for a given asset pair
  const orderbook: Book = {
    buyTree: null,
    sellTree: null,
    lowestSell: null,
    highestBuy: null,
    orderMap: {},
    limitMap: {},
  };

  const addOrder = (order: OrderSubmissionInterface) => {
    const orderId = String(Number(getLastOrderId(orderbook)) + 1); // Will lose precision at 2^53 orders
    const newOrder: Order = {
      idNumber: orderId, // TODO: Implement monotonically increasing order IDs
      buyOrSell: order.buyOrSell,
      shares: order.shares,
      limit: order.limit,
      entryTime: null,
      eventTime: BigInt(Date.now()),
      nextOrder: null,
      prevOrder: null,
      parentLimit: null,
    };
    // Add the order to the orderbook.
    addOrderToOrderbook(orderbook, newOrder);
  };

  const executeOrder = (order: Order) => {};

  const cancelOrder = (order: Order) => {};

  const getOrdersAtLimitPrice = (limitPrice: bigint, buy: boolean) => {
    const limit = findClosestLimitFromPrice(
      limitPrice,
      buy ? orderbook.buyTree : orderbook.sellTree
    );
    if (limit === null) {
      return;
    }
    return limit.headOrder;
  };

  function getAllOrders() {
    return orderbook.orderMap;
  }

  function getAllLimits() {
    return orderbook.limitMap;
  }

  // Return the functions required to interact with the orderbook.
  return {
    addOrder,
    cancelOrder,
    executeOrder,
    getOrdersAtLimitPrice,
    getAllOrders,
    getAllLimits,
  };
};
