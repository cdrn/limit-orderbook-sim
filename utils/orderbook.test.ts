import { assert } from "node:console";
import { FastLimitOrderbook } from "./orderbook";

// Mock away Date.now such that our timestamps don't break snapshots
jest
  .spyOn(global.Date, "now")
  .mockImplementation(() => new Date("2019-05-14T11:01:58.135Z").valueOf());

const generateRandomOrders = (numOrders: number) => {
  const orders = [];
  for (let i = 0; i < numOrders; i++) {
    orders.push({
      buyOrSell: Math.random() > 0.5 ? true : false,
      limit: BigInt(Math.floor(Math.random() * 100)),
      shares: BigInt(Math.floor(Math.random() * 100)),
    });
  }
  return orders;
};

test("a Fastlimit orderbook object can be instantiated", () => {
  assert(FastLimitOrderbook());
});

test("An order can be added to the orderbook correctly", () => {
  const orderbook = FastLimitOrderbook();
  const order = {
    buyOrSell: true,
    shares: BigInt(1),
    limit: BigInt(1),
    entryTime: BigInt(1),
    eventTime: BigInt(1),
    nextOrder: null,
    prevOrder: null,
    parentLimit: null,
  };
  orderbook.addOrder(order);
  const limitOrder = orderbook.getOrdersAtLimitPrice(BigInt(1), true);
  expect(limitOrder).toMatchSnapshot();
});

test("Add multiple orders at different limits to the orderbook", () => {
  const orderbook = FastLimitOrderbook();
  const orders = generateRandomOrders(500);

  console.log(orders.length, "orders generated!");

  for (const order of orders) {
    orderbook.addOrder(order);
  }
  const limitBuyOrders = orderbook.getOrdersAtLimitPrice(BigInt(1), true);
  console.log("LIMITS");
  console.log(orderbook.getAllLimits());
  console.log("ORDERS LENGTH");
  console.log(orderbook.getAllOrders());
});
