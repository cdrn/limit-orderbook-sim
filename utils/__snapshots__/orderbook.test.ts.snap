// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An order can be added to the orderbook correctly 1`] = `
Object {
  "buyOrSell": true,
  "entryTime": null,
  "eventTime": 1557831718135n,
  "idNumber": "1",
  "limit": 1n,
  "nextOrder": null,
  "parentLimit": null,
  "prevOrder": null,
  "shares": 1n,
}
`;
